### DISCUSSION TITLE
network/networking

### DISCUSSION THREAD

**Discussion by giscus** [isno/theByteBook#174] [General]
# network/networking

构建大规模高可用的分布式系统

https://www.thebyte.com.cn/network/networking.html

**Comment by xdsdmg** [comment]
图 3-2 可能有问题（不是十分确定），印象里 L2 是由网卡驱动实现的，内核负责实现 L3、L4。

**Reply by isno** [reply]
L2 是内核实现的啊，网卡驱动只处理设备、ringbuffer 相关的工作。不然 MAC 地址解析、ARP 协议怎么做呢？

**Reply by isno** [reply]
那个框里面的文字 “Kernel Stack” 不太合适，我想换成 Kernel Space 更合适一点。

**Reply by xdsdmg** [reply]
查了下资料，ARP 协议确实是由 Linux 内核实现的 [IPv4 ARP 实现源码](https://github.com/torvalds/linux/blob/master/net/ipv4/arp.c)，但是 ARP 协议属于 L2 还是 L3 是有争议的 [参考链接](https://en.wikipedia.org/wiki/Address_Resolution_Protocol#Layering)，但是 Wi-Fi、Ethernet 等数据链路层协议应该是由驱动实现的。

所以严格来讲，在 L2，ARP 协议（但是否属于 L2 有争议）由内核实现，数据传输的相关协议是由驱动实现的。

**Reply by isno** [reply]
谢谢指正，我这写得不严谨，配图也有错误，已更新修正。

**Reply by xdsdmg** [reply]
没事儿没事儿，文章写得很棒，学到了很多，工作摸鱼的时候经常看hh

**Reply by adcen0107** [reply]
不能这么说，L2有很多协议，大部分L2工作都是由网卡完成的，少部分是由内核完成的。所以你们想明确定义全部由网卡或者内核完成都是不对的。

**Comment by littlebear-xbz** [comment]
无用是高并发的挑战，还是容器网络中 overlay/underlay 的设计选型
文章倒数第二段，这个**无用**是不是应该是 **无论**

**Reply by isno** [reply]
谢谢，已修复。

**Comment by DAWEIJSW** [comment]
请问这个是用什么画的框图？

**Reply by isno** [reply]
Google 绘图：https://docs.google.com/drawings/

**Comment by wudilun123** [comment]
传输层（L3 Network layer）=> 传输层（L4 Network layer）

**Reply by isno** [reply]
谢谢，已修改
